snuba/pipeline/query_pipeline.py

```diff
@@ -76,3 +81,51 @@ def build_planner(
         self, query: Query, settings: QuerySettings
     ) -> QueryPlanner[ClickhouseQueryPlan]:
         raise NotImplementedError
+
+
+class QueryPipelineStage(Generic[Tin, Tout]):
+    """
+    This class represents a single stage in the snuba query execution pipeline.
+    The purpose of this class is to provide an organized and transparent interface to
+    execute specific processing steps on the query with clearly defined inputs and outputs.
+    These stages are designed to be composed and/or swapped among each other to form a
+    a flexible query pipeline.
+
+    Some examples of a query pipeline stage may include:
+    * Execute all entity query processors defined on the entity yaml
+    * Apply query transformation from logical representation to Clickhouse representation
+    * Execute all storage processors defined on the storage yaml
+    * Run query execution
+    * Query reporting
+
+    To create a Query Pipeline Stage, the main components to specify are:
+    an input type, an output type, and a execution function which returns the output wrapped with QueryPipelineResult.
+    ==============================================
+        >>> class MyQueryPipelineStage(QueryPipelineStage[LogicalQuery, LogicalQuery]):
+        >>>    def _execute(self, input: QueryPipelineResult[LogicalQuery]) -> QueryPipelineResult[LogicalQuery]:
+        >>>         try:
+        >>>             result = my_complex_processing_function(input.data)
+        >>>             return QueryPipelineResult(result, None)
+        >>>         except Exception as e:
+        >>>             return QueryPipelineResult(None, e)
+    """
+
+    @abstractmethod
+    def _execute(self, input: QueryPipelineResult[Tin]) -> QueryPipelineResult[Tout]:
+        raise NotImplementedError
+
+    def execute(self, input: QueryPipelineResult[Tin]) -> QueryPipelineResult[Tout]:
+        if input.error:
+            # Forward the error to next stage of pipeline
+            return QueryPipelineResult(None, input.error)
+        return self._execute(input)
+
+
+@dataclass
+class QueryPipelineResult(Generic[Tout]):
+    """
+    A container to represent the result of a query pipeline stage.
+    """
+
+    data: Optional[Tout]
+    error: Optional[Exception]
```

Comment:
Is there any tradeoff between this design and an alternative design that uses subclassing: 

class QueryPipelineResult  # abstract
class QueryPipelineData is a QueryPipelineResult
class QueryPipelineError is a QueryPipelineError

another thought is execute could return either a QueryPipelineResult or a QueryPipelineError

Reply:
I would maybe put a `__post_init__` here to make sure at least data OR error is set. They can't both be None. You could also use a `Union` type to denote that it can be either

Reply:
Subclassing can also give us some flexibility if we wanted to add different attributes or methods to the result (serialization?). Since there is no real use-case for that at the moment, we could evolve this as see what it looks like in practice. Added type validator.